### PR TITLE
Use sequence length axis in tensor trim

### PR DESCRIPTION
**Pull request by as-suvorov**
* Use sequence length axis in `trimm_tensor`

### CODE REVIEW

**Review by ilya-lavrenov — APPROVED** [review]

File: samples/cpp/prompt_lookup_decoding_lm/CMakeLists.txt
Comment (ilya-lavrenov) [inline samples/cpp/prompt_lookup_decoding_lm/CMakeLists.txt:1]: should we add such models to GHA CI validation ?

File: samples/cpp/prompt_lookup_decoding_lm/CMakeLists.txt [hunk 1]
```diff
@@ -6,16 +6,29 @@ if(TARGET openvino_tokenizers)
 else()
     message(FATAL_ERROR "multinomial_causal_lm must be compiled as part of OpenVIINOGenAI to have the path to openvino_tokenizers hardcoded.")
 endif()
+
+include(FetchContent)
+
+FetchContent_Declare(nlohmann_json
+    URL https://github.com/nlohmann/json/archive/refs/tags/v3.11.3.tar.gz
+    URL_HASH SHA256=0d8ef5af7f9794e3263480193c491549b2ba6cc74bb018906202ada498a79406)
+FetchContent_MakeAvailable(nlohmann_json)
```
Comment (ilya-lavrenov) [inline samples/cpp/prompt_lookup_decoding_lm/CMakeLists.txt:15]: please, add the similar guards for nlohmann as in https://github.com/openvinotoolkit/openvino.genai/pull/726 and define CMP policy 135

File: samples/cpp/prompt_lookup_decoding_lm/CMakeLists.txt
Comment (as-suvorov) [reply samples/cpp/prompt_lookup_decoding_lm/CMakeLists.txt:1]: Would we keep stateful mode after CB merge? If yes, then I think it make sense to add tests.

Comment (Wovchena) [reply samples/cpp/prompt_lookup_decoding_lm/CMakeLists.txt:1]: Yes, CB is already merged and we keep stateful mode.

**Review by pavel-esir — APPROVED** [review]

**Review by ilya-lavrenov — APPROVED** [review]

File: samples/cpp/prompt_lookup_decoding_lm/CMakeLists.txt
Comment (ilya-lavrenov) [reply samples/cpp/prompt_lookup_decoding_lm/CMakeLists.txt:1]: Tests can be added separately

File: samples/cpp/prompt_lookup_decoding_lm/prompt_lookup_decoding_lm.cpp [hunk 2]
```diff
@@ -1,17 +1,50 @@
...
+    if (!data.contains("model_type")) {
+        return DEFAILT_SEQ_LEN_AXIS;
+    }
+
+    const std::string model_type = data["model_type"].get<std::string>();
+
+    const std::map<std::string, size_t> model_type_to_seq_len_axis{
+        {"chatglm", 0},
+        {"llama", 2},
+    };
```
Comment (as-suvorov) [inline samples/cpp/prompt_lookup_decoding_lm/prompt_lookup_decoding_lm.cpp:40]: @ilya-lavrenov , @Wovchena , This approach is not reliable, [chatglm3](https://huggingface.co/THUDM/chatglm3-6b/blob/main/config.json) and chatglm2 have `seq_len_axis=0` but [glm-4-9b-chat](https://huggingface.co/THUDM/glm-4-9b-chat/blob/main/config.json) has `seq_len_axis=2`. All of them have same `model_type='chatglm'`.
I'll try to detect `seq_len_axis` by comparing kv tensor shape dimensions with generated tokens length.

**Comment by as-suvorov** [comment]
Removed from merge queue due to: https://github.com/openvinotoolkit/openvino.genai/pull/723#discussion_r1701747692

**Review by ilya-lavrenov — APPROVED** [review]

File: .github/workflows/causal_lm_cpp.yml [hunk 3]
```diff
@@ -420,7 +420,7 @@ jobs:
           A:' > ./prompt.txt
 
           ./build/samples/cpp/prompt_lookup_decoding_lm/prompt_lookup_decoding_lm ./TinyLlama-1.1B-Chat-v1.0/ "$(<prompt.txt)" > predictions_prompt_lookup.txt
-          ./build/samples/cpp/prompt_lookup_decoding_lm/prompt_lookup_decoding_lm ./TinyLlama-1.1B-Chat-v1.0/ "$(<prompt.txt)" > predictions_greedy.txt
+          ./build/samples/cpp/greedy_causal_lm/greedy_causal_lm ./TinyLlama-1.1B-Chat-v1.0/ "$(<prompt.txt)" > predictions_greedy.txt
```
Comment (Wovchena) [inline .github/workflows/causal_lm_cpp.yml:423]: It seems it should be reverted

Comment (ilya-lavrenov) [reply .github/workflows/causal_lm_cpp.yml:423]: why? references should be generated w/o speculative decoding, IMO

Comment (Wovchena) [reply .github/workflows/causal_lm_cpp.yml:423]: You are correct

**Review by Wovchena — APPROVED** [review]

**Pull request merged**